### ISSUE TITLE
Installer-related classes missing

### ISSUE THREAD

**Issue by cbenard** [dotnet/runtime#29643]
ServiceController was added to corefx, but many framework classes are lacking. With the deprecation of .NET Framework, I've been looking at Portability Analyzer and one of our significant issues is installer-related. We have a custom base service DLL we use in all of our services that handle installation of various things (event log sources, the assemblies themselves, service registration, etc) as well as uninstallation.

It uses the following classes which are missing. I haven't been able to find them in compat packs, NuGet, or corefx. Also, even looking for `ServiceProcessInstaller`, it doesn't appear to be in the reference source. Its signature is there, like an interface, but not the implementation.

Can anyone provide guidance on replacing the following classes missing from .NET Core/5?

* System.Diagnostics.EventLogInstaller
* System.Configuration.Install.InstallContext
* System.ServiceProcess.ServiceInstaller
* System.Configuration.Install.Installer
* System.ServiceProcess.ServiceProcessInstaller
* System.Configuration.Install.InstallerCollection
* System.Configuration.Install.TransactedInstaller
* System.Configuration.Install.InstallException
* System.ServiceProcess.ServiceAccount
* System.Configuration.Install.UninstallAction

The examples with Windows Services they now support have shown MS staff using `sc` to create the services manually (presumably xcopying the binaries manually too). This seems weird since there was built in support for registering services before and Core now officially supports services. Why are these classes still missing?

**Comment by Misiu** [comment]
`ServiceInstaller` would be very useful.
When setting the `CanHandlePreshutdownEvent` [will be added](https://github.com/dotnet/runtime/issues/30386) to `ServiceBase` then we need a way to install the service and set Timeout.

**Comment by cbenard** [comment]
@Misiu and others, I went ahead and created [Core.System.ServiceProcess](https://github.com/cbenard/Core.System.ServiceProcess). It's in [NuGet](https://www.nuget.org/packages/Core.System.ServiceProcess/) too.

It relies on @flamencist's [Core.System.Configuration.Install](https://github.com/flamencist/Core.System.Configuration.Install).

**Comment by cbenard** [comment]
@danmosemsft sorry to bother you again, but `System.Configuration.Install` was never added to the reference source. As such @flamencist's is based on decompilation of the 4.0 assemblies, not the reference source. Could we please get the official source?

**Comment by danmosemsft** [comment]
@cbenard does this look like everything? https://github.com/microsoft/referencesource/pull/145

**Comment by cbenard** [comment]
@danmosemsft I can only really compare to [Core.System.Configuration.Install](https://github.com/flamencist/Core.System.Configuration.Install/tree/master/System.Configuration.Install) from 
@flamencist since that's what I'm using now. He might be able to provide some more context.

There do appear to be some file differences but I'm not sure if that's what he had to do via decompilation.

**Comment by paul-datatech911** [comment]
Shouldn't the ServiceInstaller & ServiceProcessInstaller classes be part of the [System.Windows.Compatibility](https://www.nuget.org/packages/Microsoft.Windows.Compatibility) packages on top of the net5/net6 runtime?

**Comment by danmoseley** [comment]
@paul-datatech911 that's presumably where they would go if we ported them. Right now this isn't planned for 7.0. @cbenard did you end up creating something useable from the reference source?

**Comment by paul-datatech911** [comment]
That's unfortunate because even the [Microsoft Installer Projects Extension](https://marketplace.visualstudio.com/items?itemName=VisualStudioClient.MicrosoftVisualStudio2022InstallerProjects) uses these class as CustomActions to easily install Services.  

I realize there are other ways out there, but that leaves a big hole in the default Installer capability shipped out of the box.

**Comment by paul-datatech911** [comment]
> 

@cbenard I was able to consume your classes from the Nuget package to fix all the build errors related to the installer classes - thank you for that!

The problem I have now is that I actually have custom actions that extend the Installer Classes that I use via the [Microsoft Installer Projects](https://marketplace.visualstudio.com/items?itemName=VisualStudioClient.MicrosoftVisualStudio2022InstallerProjects), and these look at thebinaries and presumably use reflection to see if they are in-fact installer classes.  When you run the installer, you get the following error message presumably because although there is an installer class present, it isn't the one it's looking for (different namespace, different assemblyToken?).  

![image](https://user-images.githubusercontent.com/78918644/149806439-f9e364e0-e8dc-40c4-ac8e-d6e13e5b6ab2.png)

![image](https://user-images.githubusercontent.com/78918644/149806702-4614bfd9-80ac-4ddd-85f7-bdf2b70960de.png)

My Installer classes generally look like this:
```[RunInstaller(true)]
public partial class FR911ApiCheckInstaller : System.Configuration.Install.Installer
{
  ...
}
```

And they use the implementation of Installer from the core.system.configuration.install package
```
#region Assembly System.Configuration.Install, Version=1.1.0.0, Culture=neutral, PublicKeyToken=null
// C:\Users\pauls\.nuget\packages\core.system.configuration.install\1.1.0\lib\netstandard2.0\System.Configuration.Install.dll
// Decompiled with ICSharpCode.Decompiler 6.1.0.5902
#endregion

using Newtonsoft.Json.Linq;
using System.Collections;
using System.ComponentModel;
using System.Linq;
using System.Text;

namespace System.Configuration.Install
{
       //
    // Summary:
    //     Provides the foundation for custom installations.
    [DefaultEvent("AfterInstall")]
    public class Installer : Component
    {
        ...
    }
}
```

I think the only way to solve this is to get the Installer classes where they belong in the [Microsoft.Windows.Compatibility](https://www.nuget.org/packages/Microsoft.Windows.Compatibility/6.0.0) package and have the Installer extension team use those to build the extension against?  Any other suggestions?  

I understand that we can use the basic functionality of sc.exe to get services installed, but we actually use custom UI in the installer project to gather configuration data at install time, and there are provisions built into the installer to pass data around from UI to installer classes via CustomActionData.

**Comment by cbenard** [comment]
@paul-datatech911 I'm not using an installer project, and I haven't looked at this project in a long time. Going forward, I'm probably not going to be using any of these deprecated namespaces and I will refactor how our in-house stuff works to not use any of the *Installer classes. It will probably just use `sc` under the hood and file copying/directory creation without using Installer subclasses.

**Comment by ericstj** [comment]
The old install-util based custom actions had two ways of running.  1) using `<frameworkDir>InstallUtil.exe` which was just a wrapper .NETFramework exe that hosted the user DLL and forwarded on command-line arguments or 2) `InstallUtilLib.dll` which exposed a native custom action entry-point to MSI that could get the database handle and fetch properties and CustomActionData and pass it to managed custom action.  

The latter (2) was tightly coupled to .NETFramework since it needed to locate the framework and start it up.  It wouldn't make sense to directly port that to .NETCore, it would need to be rewritten with knowledge of how to find and host .NETCore.  It would need to handle isolation, versioning, etc.  Even if all that was done I don't think the folks maintaining VDProj would want to touch it to make it aware of some new type of managed custom action.  The error you show above "Unable to get installer types" is because you have a .NETFramework host component looking in your .NETCore assembly for Installer derived types, and it won't find them.  Even if it did, you wouldn't want a .NETFramework component running the install operation for your .NETCore service (at least I don't think you would).

The former (1) doesn't really make sense in the context of .NETCore: there's not a "known single install point" for .NETCore.  It versions rather quickly side-by-side.  So having an installer shell out to some known location for a framework EXE to then point it at your user assembly is pushing significant complexity (find the framework) into the installer.  A better solution would be to bundle an exe with your app and call into that.  If you wanted, it could be a .NETCore application which leveraged runtimeconfig settings to locate the framework it runs on.  You could pass whatever data you want on the command-line.

**Comment by paul-datatech911** [comment]
> [Microsoft.Windows.Compatibility](https://www.nuget.org/packages/Microsoft.Windows.Compatibility/6.0.0)

Isn't that what [Microsoft.Windows.Compatibility ](https://www.nuget.org/packages/Microsoft.Windows.Compatibility/6.0.0) is for?  

Even if you don't believe so, there's a mismatch between what the Microsoft Visual Studio Installer Projects can see as InstallClasses and what they can actually do.

**Comment by ericstj** [comment]
> Isn't that what Microsoft.Windows.Compatibility is for?

`Microsoft.Windows.Compatibility` is a collection of all the packages that have types that were in `.NETFramework` that we also ported.  If we had ported all of `System.Configration.Install` it would be included, however we did not port it for the reasons mentioned here and in the linked discussions.

> there's a mismatch between what the Microsoft Visual Studio Installer Projects can see as InstallClasses and what they can actually do.

If my memory serves, VDProj is just checking if you gave it a .NET assembly, it's not giving you a guarantee that your assembly contains types that load and execute.  At runtime the custom action host (installutillib) just starts up the .NETFramework runtime using [ClrCreateManagedInstance](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/hosting/clrcreatemanagedinstance-function) to create a [ManagedInstallerClass](https://docs.microsoft.com/en-us/dotnet/api/system.configuration.install.managedinstallerclass?view=netframework-4.8).  It then invokes the [ManagedInstall method](https://docs.microsoft.com/en-us/dotnet/api/system.configuration.install.managedinstallerclass.system-configuration-install-imanagedinstaller-managedinstall?view=netframework-4.8#system-configuration-install-managedinstallerclass-system-configuration-install-imanagedinstaller-managedinstall(system-string-system-int32)) ([reference source](https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.Configuration.Install/ManagedInstaller.cs#L34)) passing in your CA's CustomActionData as `argString` and the custom action's `MSIHANDLE` as `hInstall`.   This then examines the types in your assembly that meet [this criteria](https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.Configuration.Install/AssemblyInstaller.cs#L312-L314).  As you suggest, it doesn't find your types because they derive from types that don't unify with the .NETFramework types.

Now, it's possible that you could build a facade assembly for `https://github.com/cbenard/Core.System.ServiceProcess/tree/master/System.ServiceProcess.Core` targeting .NETFramework with typeforwards to `System.ServiceProcess` and find some way to sim-ship those with the .NETCore versions such that your service loaded the .NETCore version but the CA loaded the .NETFramework version, but you'd still be executing in .NETFramework because that's what VDProj and `InstallUtilLib` do.  So you'd be going to a lot of effort just to get something that'd be a lot easier to get by just keeping your installer types targeting .NETFramework.

TLDR - InstallUtilLib dll based custom actions are tightly tied to .NETFramework.

If you want to run your custom action in .NETCore then your best option is to run EXE custom actions.  @cbenard's project gives one option for that, with similar code to InstallUtil / System.Configuration.Install.  Alternatively you could write your own app that has whatever behavior you like (as @cbenard suggests he's doing now).  For sharing data with your installer you can pass that via the commandline.  Though the old InstallUtil stuff accepted an MSIHandle, it never shared it with the running installer and the only thing it used that for was an [error string](https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System.Configuration.Install/ManagedInstaller.cs#L55) which is a rather small loss when switching to an EXE custom action.

Hope this helps!
Eric

**Comment by cbenard** [comment]
To be clear, like @ericstj said, we never were using `installutil.exe`. We have a `-install` and `-uninstall` flag that internally runs our `Installer` classes. Similar to Top Shelf, as I understand it. What I did was never intended to allow `installutil` to be used under Core.

Thanks for all the extra info/context @ericstj.

**Comment by danmoseley** [comment]
@cbenard thanks again for putting up the `Core.System.ServiceProcess` package. I've had other customers be successful with it and I see you've got healthy downloads.

I'll close this now as there's no plans to port our own. Hopefully others will discover the package through here or https://github.com/cbenard/Core.System.ServiceProcess.
